pkg/build/gobuild.go

```diff
@@ -755,31 +813,18 @@ func (g *gobuild) Build(ctx context.Context, s string) (Result, error) {
 	if err != nil {
 		return nil, err
 	}
-
-	// Annotate the image or index with base image information.
-	// (Docker manifest lists don't support annotations)
-	if mt != types.DockerManifestList {
-		anns := map[string]string{
-			specsv1.AnnotationBaseImageDigest: baseDigest.String(),
-		}
-		if _, ok := baseRef.(name.Tag); ok {
-			anns[specsv1.AnnotationBaseImageName] = baseRef.Name()
-		}
-		res = mutate.Annotations(res, anns).(Result)
-	}
-
 	return res, nil
 }
 
 // TODO(#192): Do these in parallel?
-func (g *gobuild) buildAll(ctx context.Context, ref string, baseIndex v1.ImageIndex) (v1.ImageIndex, error) {
+func (g *gobuild) buildAll(ctx context.Context, ref string, baseIndex v1.ImageIndex) (oci.SignedImageIndex, error) {
 	im, err := baseIndex.IndexManifest()
 	if err != nil {
 		return nil, err
 	}
 
 	// Build an image for each child from the base and append it to a new index to produce the result.
-	adds := []mutate.IndexAddendum{}
+	adds := []ocimutate.IndexAddendum{}
```

Comment:
This feels a bit like a smell, that we have to use cosign's mutate methods instead of ggcr's, throughout this code. And that `buildOne` only ever returns a (misnomered) `SignedImage`.

AIUI the `Signed*` constructs exist so that we can attach things to it, can't cosign's packages just take a regular `v1.Image` or `v1.ImageIndex` in its `Attach*` methods?

(I'm fine with forging ahead despite this smell, but I'd love to have a better way to compose these things than having `ko` -- and any other tool that wants to integrate attachment -- adopt `Signed*` variants everywhere)

Reply:
> I'd love to have a better way to compose these things than having ko -- and any other tool that wants to integrate attachment -- adopt Signed* variants everywhere

I'm certainly open to alternatives, but I would expect some degree of this unless/until OCI itself has constructs for attachments, signatures, attestations.  Once that happens, GGCR gets API for it, and the Cosign-specific aspects die off.

Until then, the whole point of the "cosign as an SDK" efforts has been to make this as small a lift for GGCR users as we can make it (in terms of api familiarity and compatibility).

Reply:
Synced up quickly on meet.  I think we're pretty well aligned that if/as Attachment-like functionality lands in GGCR that a lot of the need for cosign to define it's own interfaces for images/indices will subside, and it will devolve into providing higher-level access patterns for building / accessing things within attachments (if attachments continue to have a multiple parts, e.g. signature-per-layer).